src/IO/ReadBufferFromS3.cpp

```diff
@@ -300,6 +301,9 @@ std::unique_ptr<ReadBuffer> ReadBufferFromS3::initialize()
     {
         read_result = outcome.GetResultWithOwnership();
 
+        if (read_settings.throttler)
+            read_settings.throttler->add(read_result.GetContentLength());
```

Comment:
Actually I want it to be as near as possible to real network operation. But I'm not sure about all these S3/HDFS/Azure clients, when they do network IO. And, yes, if the whole file is the only possible place for throttling, it will be better to use readImpl, just to avoid stopping all IO after one big add(), I'll fix here.